hack/upgrade-test.sh

```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/env bash
+# Copyright 2021 VMware
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# shellcheck disable=SC2155
+readonly DIR="$(cd "$(dirname "$0")" && pwd)"
+readonly HOST_ADDR=${HOST_ADDR:-$("$DIR"/ip.py)}
+readonly REGISTRY_PORT=${REGISTRY_PORT:-5000}
+readonly REGISTRY=${REGISTRY:-"${HOST_ADDR}:${REGISTRY_PORT}"}
+
+readonly TEST_NAME="gitwriter-sc"
+readonly GIT_SERVER="git-server.default.svc.cluster.local:80"
+readonly SOURCE_REPO="hello-world"
+readonly SOURCE_BRANCH="main"
+readonly CONFIG_REPO="gitops-test"
+readonly CONFIG_BRANCH="main"
+readonly CONFIG_COMMIT_MESSAGE="Update config"
+
+main() {
+  ./hack/setup.sh cluster cartographer-latest example-dependencies
+
+  port=$(available_port)
+  setup_git_server "$port"
+
+  source_dir="$(mktemp -d)"
+  setup_source_repo "$source_dir" "$port"
+  setup_source_to_gitops
+
+  config_dir="$(mktemp -d)"
+  test_source_to_gitops "$config_dir" "$port"
+
+  setup_gitops_to_app
+  test_gitops_to_app
+
+  log "deploying cartographer from main"
+  ./hack/setup.sh cartographer
+  verify_new_carto_deployed
+
+  update_source "$source_dir" "$port"
+  wait_for_new_config_commit "$config_dir"
+  wait_for_knative_deployment_update
+}
+
+setup_git_server() {
+  local port=$1
+
+  log "setting up git server"
+
+  kubectl apply -f ./hack/git-server.yaml
+
+  until [[ $(kubectl get deployment git-server -o json | jq '.status.readyReplicas') == 1 ]]; do
+    log "waiting for git-server deployment"
+    sleep 10
+  done
+
+  (kubectl port-forward service/git-server $port:80 &) >/dev/null
```

Comment:
I think this will leak port forwarding process behind 🤔  

when doing something similar, I remember going for the following

```
                kubectl -n $namespace port-forward service/$name $port:80 &
                trap "kill $! || true" EXIT
```

that `trap` there gives us the ability of having that `kill <pid>` going on regardless of how we exit the process (sure, a SIGKILL on bash would prevent it 😅  but regular terminations would all get the kill going)